activerecord/lib/active_record/relation/spawn_methods.rb

```diff
@@ -37,7 +38,7 @@ def merge(other)
     end
 
     def merge!(other) # :nodoc:
-      if !other.is_a?(Relation) && other.respond_to?(:to_proc)
+      if other.is_a?(Proc)
```

Comment:
I wonder whether anyone has use the duck typing of responding to `to_proc` in their apps - we could reorder the conditionals like this to avoid triggering `to_proc` on the `Hash`:

``` ruby
def merge!(other) # :nodoc:
  if other.is_a?(Hash)
    Relation::HashMerger.new(self, other).merge
  elsif other.respond_to?(:to_proc)
    instance_exec(&other)
  else
    Relation::Merger.new(self, other).merge
  end
end
```

@rafaelfranca @sgrif wdyt?